src/Umbraco.Web/Unversion/UnversionService.cs

```diff
@@ -0,0 +1,169 @@
+
+                    // no need to compare dates & maxcount since we've already checked this version and we want to keep it
+                    continue;
+                }
+
+                // If we have a maxCount and the current iteration is MORE that max-count
+                if (configEntry.MaxCount > 0 && iterationCount > configEntry.MaxCount)
+                {
+                    _logger.Debug<UnversionService>("Mark this version to be removed {VersionId} for {ContentUdi} because iterationCount is {IterationCount} and max count is {MaxCount}", version.VersionId, version.GetUdi(), iterationCount, configEntry.MaxCount);
+                    versionIdsToDelete.Add(version.VersionId);
+
+                    // no need to compare dates since we've already added this version for deletion
+                    continue;
+                }
+
+                // If we have a max days and the current version is older
+                if (configEntry.MaxDays > 0 && configEntry.MaxDays != int.MaxValue)
+                {
+                    var dateRemoveBefore = currentDateTime.AddDays(0 - configEntry.MaxDays);
+                    if (version.UpdateDate < dateRemoveBefore)
+                    {
+                        _logger.Debug<UnversionService>("Mark this version to be removed {VersionId} for {ContentUdi} because version update date {UpdateDate} is less than the cuttoff date {DateRemoveBefore}", version.VersionId, version.GetUdi(), version.UpdateDate, dateRemoveBefore);
+                        versionIdsToDelete.Add(version.VersionId);
+
+                        // Added to list stop carrying on
+                        continue;
+                    }
+                }
+
+                // We have a mincount set and the iteration count is higher than the versions we want to keep
+                // AND we do NOT have MaxCount or MaxDays sets
+                // Thus we need to remove all other versions as we explicity only wanted to keep x min versions only
+                if(configEntry.MinCount > 0 && configEntry.MaxCount !> 0 && configEntry.MaxDays !> 0)
+                {
+                    _logger.Debug<UnversionService>("Mark this version to be removed {VersionId} for {ContentUdi} because we have all the versions we required for mincount and MaxDays nor MaxCount has been set in conjuction.", version.VersionId, version.GetUdi(), iterationCount, configEntry.MaxCount);
+                    versionIdsToDelete.Add(version.VersionId);
+                }
+            }
+
+            return versionIdsToDelete;
+
+        }
+
+        /// <summary>
+        /// Gets a list of int NodeIds from an xPath query
+        /// </summary>
+        private List<int> GetNodeIdsFromXpath(string xpath)
```

Comment:
This is returning nodes from the cache? Why? What about unpublished items? We should absolutely not be relying on the front-end cache for this processing. IMO we shouldn't be doing any XPath lookups and only relying on doc types.